keps/sig-instrumentation/4785-crdmetrics-controller/README.md

```diff
@@ -0,0 +1,1061 @@
+Resources' schemata. This was a highly appreciated and much needed feature-set,
+since before this, the only way to generate metrics for a particular resource
+was to get that logic merged upstream. It has been two years since the patch
+was merged, and yet even today the feature-set is arguably one of the most
+contributed-to parts of one of the most active repositories that fall under SIG 
+Instrumentation.
+
+However, during the recent releases, the maintainers realized that owing to the
+almost zero dependency parsing logic, it became more prone to side effects as
+the configuration scaled to include more fields to cater to the needs raised by
+the community. After many cycles of trading off maintenance complexity for more
+features, and seeing its maturity, the maintainers agreed on putting this on a 
+"maintenance-only" mode in favor of a better solution that prioritizes 
+scalability in the long run, while ensuring that it meets the same expectations
+as its predecessor and more.
+
+Such a "solution" should allow the maintainers to deprecate and drop the Custom
+Resource State API from Kube State Metrics, and replace it by the CRDMetrics 
+controller which, in addition to its own benefits, would allow Kube State 
+Metrics to drop all Custom Resource State API-specific behaviors that can crash
+Kube State Metrics, directly affecting the availability of native metrics 
+defined in the codebase, and ensuring that native metrics do not experience
+significant downtime due to an unavoidable error during the generation of an
+entirely different set of metrics. Also, clusters requiring only native metrics
+will not experience the additional Custom Resource State API overhead
+needlessly when metrics stores' are being built.
+
+The presence of native and custom resource metrics under the same hood also 
+encouraged folks to infer implicit expectations that we do not officially 
+support, which in turn have driven them to patch the two, sharing logic, when
+both anticipate varying degrees of engagement and should not be (and are not
+designed to be) interdependent. Kube State Metrics, owing to its original
+purpose of native metrics generation, is meant to be comparatively much more
+stable, and as such, the two feature-sets should be able to grow and release
+independently in contrast to a part of the codebase that the other is not
+dependent upon warranting a release.
+
+[kubernetes/kube-state-metrics#1710]: https://github.com/kubernetes/kube-state-metrics/pull/1710
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+The KEP targets a controller which:
+* allows for Custom Resource metrics generation based on their schemata,
+* while providing cluster-scoped managed resources (`CRDMetricsResource`) that 
+  allows defining the collection configuration for generating metrics 
+  on-the-fly, and,
```

Comment:
🎉  This will be awesome. Having this as a CRD also makes the configuration versioned 👍 

would be good to keep the API package which defines the type clean so importing it in other projects (e.g. https://github.com/kubernetes-sigs/controller-tools ) would not introduce other transitive dependencies.

I plan to add a generator for this in controller-tools so the config can be generated from markers just like the CRD itself.

Prior art regarding transitive dependencies for an API package:

https://github.com/kubernetes-sigs/cluster-api/issues/9011

Reply:
Thank you for the interest @chrischdi, appreciate it!

Yep, that's a valid concern. I just glanced through [`types.go`](https://github.com/rexagod/crdmetrics/blob/main/pkg/apis/crdmetrics/v1alpha1/types.go) and the imports seem nothing other than what we'd usually expect, so I think we should be good on that front.